designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+
+> It’s only in the context of trying to assign a density on the output value that a Jacobian is needed, so the proper encapsulation is
+
+```stan
+functions {
+ real f(real x) {
+   // transform function
+ }
+ real jacobian(real x) {
+   // jacobian determinant calculation
+ }
+ real prob(real x) {
+   // lpdf calculation
+ }
+}
+parameters {
+ real theta;
+}
+transformed parameters {
+ real eta = f(theta);
+}
+model {
+ // In a pure sense, neither line is well defined on it's
+ //  own, but the combinations are well defined.
+ eta ~ prob(values);
+ target += jacobian(theta);
+}
+```
+
+So users have to choose between either leaving the nice constrained space to write your model in the unconstrained space or to restrict the models developed on constrained spaces to just the ones available with the Stan languages constrained types. The hands off or hands on approach also leaves out a growing group of Stan programmers who write code for other Stan programmers. In addition, adding a constraint to Stan involves having to touch the math library which is rather complicated. Not just by C++ itself, but the essentially small DSL we use to describe autodiff in the Stan math library.
+
+Having a new block for user-defined constraints allows users and developers to remove some of the veil of encapsulation and write repeatable code for future models.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following section is a draft of the docs for the new constraint and `jacobian` keyword.
+
+## Constraint Block
+
+### Overview
+
+The constraints block was included in Stan 2.XX, introduced to provide users with the ability to define custom constraints on parameters, transformed parameters, and data. This block is situated immediately after the `functions` block and before the `data` block in a Stan program. The primary purpose of the `constraints` block is to allow users to define custom constraints that can be used in the same way as built-in constraints.
+
+### Syntax
+
+A constraints block is defined by the keyword `constraints`. Inside of the `constraints` block each constraint is its own block with a given `constraint_name` and the functions needed to define the constraint. Inside the `constraints` block, users can define one or more constraint types. Each constraint type is encapsulated within its own set of curly braces and is identified by a unique name.
+
+Each user-defined constraint within the constraint block must specify a set of functions that define the transformation between unconstrained and constrained spaces, validate data, and increment the log of the absolute value of the determinant of the Jacobian. The general structure of a user-defined constraint is as follows:
+
+```stan
```

Comment:
I think you want to describe the signatures and then give a complete example in the doc.

Reply:
I moved the paragraph below the example up which I think makes this more clear